File: app/src/main/java/com/dublikunt/nclient/components/ThreadAsyncTask.java
package com.dublikunt.nclient.components;

import androidx.appcompat.app.AppCompatActivity;

public abstract class ThreadAsyncTask<Params, Progress, Result> {
    private final AppCompatActivity activity;
    private Thread thread;

    public ThreadAsyncTask(AppCompatActivity activity) {
        this.activity = activity;
    }

    public void execute(Params... params) {
        thread = new AsyncThread(params);
        thread.start();
    }

    protected void onPreExecute() {
    }

    protected void onPostExecute(Result result) {
    }

    protected void onProgressUpdate(Progress... values) {
    }

    protected abstract Result doInBackground(Params... params);

    protected final void publishProgress(Progress... values) {
        if (!activity.isDestroyed())
            activity.runOnUiThread(() -> onProgressUpdate(values));
    }

    class AsyncThread extends Thread {
        Params[] params;

        AsyncThread(Params[] params) {
            this.params = params;
        }

        @Override
        public void run() {
            if (!activity.isDestroyed())
                activity.runOnUiThread(ThreadAsyncTask.this::onPreExecute);
            Result result = doInBackground(params);
            if (!activity.isDestroyed())
                activity.runOnUiThread(() -> onPostExecute(result));
        }
    }
}

File: app/src/main/kotlin/com/dublikunt/nclient/components/ThreadAsyncTask.kt
package com.dublikunt.nclient.components

import androidx.appcompat.app.AppCompatActivity
import com.dublikunt.nclient.settings.Global

abstract class ThreadAsyncTask<Params, Progress, Result>(private val activity: AppCompatActivity) {
    @SafeVarargs
    fun execute(vararg params: Params) {
        val thread: Thread = AsyncThread(params as Array<Params>)
        thread.start()
    }

    protected fun onPreExecute() {}
    protected open fun onPostExecute(result: Result) {}
    protected open fun onProgressUpdate(vararg values: Progress) {}
    protected abstract fun doInBackground(vararg params: Params): Result
    @SafeVarargs
    protected fun publishProgress(vararg values: Progress) {
        if (!Global.isDestroyed(activity)) activity.runOnUiThread { onProgressUpdate(*values) }
    }

    internal inner class AsyncThread(var params: Array<Params>) : Thread() {
        override fun run() {
            if (!Global.isDestroyed(activity)) activity.runOnUiThread { onPreExecute() }
            val result = doInBackground(*params)
            if (!Global.isDestroyed(activity)) activity.runOnUiThread { onPostExecute(result) }
        }
    }
}
